src/osd/OSD.cc

```diff
@@ -5619,8 +5619,17 @@ bool OSD::_is_healthy()
   }
 
   if (is_waiting_for_healthy()) {
+     utime_t now = ceph_clock_now();
+     utime_t grace = utime_t(cct->_conf->osd_max_markdown_period, 0);
+     while (!osd_markdown_log.empty() &&
+             osd_markdown_log.front() + grace < now)
+       osd_markdown_log.pop_front();
+     if (osd_markdown_log.size() <= 1) {
+       dout(5) << __func__ << " first time marked as down,"
+               << " try reboot unconditionally" << dendl;
+       return true;
+    }
```

Comment:
I understand that this block fixes your test case, but it really looks to me like you're optimizing for a test but missing the test cases where an immediate attempt to rejoin is bad, like when the network actually isn't working properly rather than being predictably manipulated by an admin.

The same applies to your comment about trying to avoid up-to-down transitions to work around our suicides when OSDs stop behaving properly...

Reply:
> I understand that this block fixes your test case, but it really looks to me like you're optimizing for a test but missing the test cases where an immediate attempt to rejoin is bad, like when the network actually isn't working properly rather than being predictably manipulated by an admin.

I don't think this change breaks the above assumption. We won't attempt to rejoin to often, e.g., causing the osd process being killing off itself. At the opposite extreme, the whole cluster seems to be losing the ability of self-healing if we don't attempt at all. Also it is not just a test case, we've encountered these heartbeat issues in our products and I am just trying to demonstrate a stable way to re-produce here.

Reply:
Okay, this part makes sense since we're still limiting the rate.